system/CodeIgniter.php

```diff
@@ -934,7 +959,11 @@ protected function display404errors(PageNotFoundException $e)
 
             $cacheConfig = new Cache();
             $this->gatherOutput($cacheConfig, $returned);
-            $this->sendResponse();
+            if (! $this->returnResponse) {
+                $this->sendResponse();
+            } else {
+                return $this->response;
+            }
 
             return;
```

Comment:
this can be changed to:

```suggestion
            if ($this->returnResponse) {
                return $this->response;
            }

            $this->sendResponse();
            return;
```